src/stats.ml

```diff
@@ -45,3 +45,117 @@ let dump () =
 let enable () =
   enabled := true;
   at_exit dump
+
+module Catapult = struct
+  type active_state = Pervasives.out_channel * Format.formatter
+
+  type state =
+    | Disabled
+    | Path of string
+    | Active of active_state
+
+  let state = ref Disabled
+
+  let dispose () =
+    match !state with
+    | Disabled -> ()
+    | Path _ -> ()
+    | Active (channel, _) ->
+      Pervasives.output_string channel "]\n";
+      Pervasives.flush channel;
+      Pervasives.close_out channel
+
+  let enable path =
+    state := Path path;
+    at_exit dispose
+
+  let fmt () =
+    match !state with
+    | Disabled -> None
+    | Active (_, fmt) ->
+      Format.pp_print_string fmt ",";
+      Some fmt
+    | Path p ->
+      let channel = Pervasives.open_out_gen [Open_append; Open_creat] 0o666 p in
+      let fmt = Format.formatter_of_out_channel channel in
+      Format.pp_print_string fmt "[";
+      state := Active (channel, fmt);
+      Some fmt
+
+  let printf format_string =
+    match fmt () with
+    | Some fmt -> Format.fprintf fmt (format_string ^^ "\n%!")
+    | None -> Format.ifprintf Format.std_formatter format_string
```

Comment:
Ideally, it would be best to just use `printf` over `Format` here. `Format` has a lot of overhead in comparison to `Print` and we don't care too much about formatting I think.

Finally, it would be nicer to have this work with a continuation like `Logs`. E.g. `printf (fun f -> printf f "...")`

Reply:
OK I'll try that! That was to have nice `pp` functions for arguments but we can do these with a plain channel as well.